// react-lifecycle-visualizer/dist/components/Log.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _constants = require('../constants');

var constants = _interopRequireWildcard(_constants);

var _actionCreators = require('../redux/actionCreators');

var ActionCreators = _interopRequireWildcard(_actionCreators);

var _LogEntries = require('./LogEntries');

var _LogEntries2 = _interopRequireDefault(_LogEntries);

var _SimpleButton = require('./SimpleButton');

var _SimpleButton2 = _interopRequireDefault(_SimpleButton);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /* global document:false */


var DelaySelector = function DelaySelector(_ref) {
  var value = _ref.value,
      onChange = _ref.onChange;
  return _react2.default.createElement(
    'select',
    { value: value, onChange: onChange },
    constants.delayValues.map(function (delay) {
      return _react2.default.createElement(
        'option',
        { value: delay, key: delay },
        delay,
        's'
      );
    })
  );
};

var Log = function (_Component) {
  _inherits(Log, _Component);

  function Log() {
    var _ref2;

    var _temp, _this, _ret;

    _classCallCheck(this, Log);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref2 = Log.__proto__ || Object.getPrototypeOf(Log)).call.apply(_ref2, [this].concat(args))), _this), _this.onKeyDown = function (evt) {
      if (evt.shiftKey) {
        // Require shift to prevent interference with scrolling
        switch (evt.code) {
          case 'ArrowUp':
            _this.props.stepLog(-1);
            break;
          case 'ArrowDown':
            _this.props.stepLog(1);
            break;
        }
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Log, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      document.addEventListener('keydown', this.onKeyDown);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      document.removeEventListener('keydown', this.onKeyDown);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      return _react2.default.createElement(
        'div',
        { className: 'log' },
        _react2.default.createElement(
          'div',
          { className: 'header' },
          _react2.default.createElement(
            'div',
            { className: 'controls' },
            _react2.default.createElement(
              'div',
              { className: 'title' },
              'Log'
            ),
            _react2.default.createElement(
              'div',
              { className: 'buttons' },
              _react2.default.createElement(_SimpleButton2.default, { value: 'clear log', onClick: function onClick() {
                  return _this2.props.clearLog();
                } }),
              ' ',
              _react2.default.createElement(
                'span',
                { className: 'emoji-button', onClick: function onClick() {
                    return _this2.props.stepLog(-1);
                  } },
                '\u23EA'
              ),
              this.props.replayTimerId === null ? _react2.default.createElement(
                'span',
                { className: 'emoji-button', onClick: function onClick() {
                    return _this2.props.startReplay();
                  } },
                '\u25B6\uFE0F'
              ) : _react2.default.createElement(
                'span',
                { className: 'emoji-button', onClick: function onClick() {
                    return _this2.props.pauseReplay();
                  } },
                '\u23F8'
              ),
              _react2.default.createElement(
                'span',
                { className: 'emoji-button', onClick: function onClick() {
                    return _this2.props.stepLog(1);
                  } },
                '\u23E9'
              )
            ),
            _react2.default.createElement(
              'div',
              null,
              'Delay:',
              _react2.default.createElement(DelaySelector, {
                value: this.props.replayTimerDelay,
                onChange: function onChange(evt) {
                  return _this2.props.setDelay(+evt.currentTarget.value);
                }
              })
            )
          ),
          _react2.default.createElement(
            'div',
            null,
            '(hover to highlight, shift-up/down to navigate)'
          )
        ),
        _react2.default.createElement(_LogEntries2.default, {
          entries: this.props.logEntries,
          highlightedIndex: this.props.highlightedIndex,
          highlight: this.props.highlight
        })
      );
    }
  }]);

  return Log;
}(_react.Component);

var mapStateToProps = function mapStateToProps(_ref3) {
  var logEntries = _ref3.logEntries,
      highlightedIndex = _ref3.highlightedIndex,
      replayTimerId = _ref3.replayTimerId,
      replayTimerDelay = _ref3.replayTimerDelay;
  return { logEntries: logEntries, highlightedIndex: highlightedIndex, replayTimerId: replayTimerId, replayTimerDelay: replayTimerDelay };
};

exports.default = (0, _reactRedux.connect)(mapStateToProps, ActionCreators, null, { storeKey: constants.reduxStoreKey })(Log);
